synapse/lib/layer.py

```diff
@@ -1468,93 +1521,170 @@ async def tryfix(lkey, buid, form):
+
+        for lkey, buid in self.layrslab.scanByPref(abrv, db=self.byprop):
+
+            await asyncio.sleep(0)
 
-        def verifyIndxKeys(propvalu, stortype, indxkeys):
+            indx = lkey[len(abrv):]
 
-            # TODO: we dont support verifying array property indexes just yet...
+            sode = self._getStorNode(buid)
+            if sode is None:
+                await tryfix(lkey, buid)
+                yield ('NoNodeForPropIndex', {'buid': s_common.ehex(buid), 'form': form, 'prop': prop, 'indx': indx})
+                continue
+
+            if prop is not None:
+                props = sode.get('props')
+                valu = props.get(prop)
+                if valu is None:
+                    await tryfix(lkey, buid)
+                    yield ('NoValuForPropIndex', {'buid': s_common.ehex(buid), 'form': form, 'prop': prop, 'indx': indx})
+                    continue
+            else:
+                valu = sode.get('valu')
+                if valu is None:
+                    await tryfix(lkey, buid)
+                    yield ('NoValuForPropIndex', {'buid': s_common.ehex(buid), 'form': form, 'prop': prop, 'indx': indx})
+                    continue
+
+            propvalu, stortype = valu
             if stortype & STOR_FLAG_ARRAY:
-                return
+                stortype = STOR_TYPE_MSGP
 
-            # NOTE: This *will* mutate indxkeys list
             try:
                 for indx in self.stortypes[stortype].indx(propvalu):
-                    indxkey = abrv + indx
-                    if indxkey not in indxkeys:
-                        yield ('NoPropKeyForIndx', {'buid': s_common.ehex(buid), 'form': form, 'prop': prop, 'indx': indx})
-                        continue
-
-                    indxkeys.remove(indxkey)
+                    if abrv + indx == lkey:
+                        break
+                else:
+                    await tryfix(lkey, buid)
+                    yield ('SpurPropKeyForIndex', {'buid': s_common.ehex(buid), 'form': form,
+                                                   'prop': prop, 'indx': indx})
 
             except IndexError:
+                await tryfix(lkey, buid)
                 yield ('NoStorTypeForProp', {'buid': s_common.ehex(buid), 'form': form, 'prop': prop,
                                              'stortype': stortype})
 
-            for indxkey in indxkeys:
-                indx = indxkey[len(abrv):]
-                yield ('SpurPropKeyForIndx', {'buid': s_common.ehex(buid), 'form': form, 'prop': prop, 'indx': indx})
+    async def verifyByPropArray(self, form, prop, autofix=None):
 
-        for lkey, buid in self.layrslab.scanByPref(abrv, db=self.byprop):
+        abrv = self.getPropAbrv(form, prop)
 
-            await asyncio.sleep(0)
+        async def tryfix(lkey, buid):
+            if autofix == 'index':
+                self.layrslab.delete(lkey, buid, db=self.byarray)
```

Comment:
Should we require the autofix value to be `None` or `index` on entry to blow up cleanly?

Reply:
sure, i'll add a check for valid autofix strategies